nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationState.java

```diff
@@ -132,27 +136,80 @@ public void setModuleInfo(ModuleInfo moduleInfo) {
         this.moduleInfo = moduleInfo;
     }
 
+    public ApplicationInfo getApplicationInfo() {
+        return applicationInfo;
+    }
+
+    public void setApplicationInfo(ApplicationInfo applicationInfo) {
+        this.applicationInfo = applicationInfo;
+    }
+
     public Set<String> getSniffers() {
         return sniffers;
     }
 
-    public void setSniffers(Set<String> sniffers) {
-        if (!sniffers.equals(this.sniffers)) {
-            this.sniffers = sniffers;
+    public void setSniffers(Collection<? extends Sniffer> sniffers) {
+        final Set<String> snifferTypes = sniffers.stream()
+                    .map(s -> Sniffer.class.cast(s))
+                    .map(Sniffer::getModuleType)
+                    .collect(toSet());
+        if (!snifferTypes.equals(this.sniffers)) {
+            this.sniffers = snifferTypes;
             this.moduleInfo = null;
             this.engineInfos = null;
         }
     }
     
-    public void copyPreviousState(ExtendedDeploymentContext context) {
-        ExtendedDeploymentContext previousDC = this.getDeploymentContext();
-        context.getAppProps().putAll(previousDC.getAppProps());
-        context.getModulePropsMap().putAll(previousDC.getModulePropsMap());
-        previousDC.getModuleMetadata().forEach(context::addModuleMetaData);
+    public void copyPreviousState(ExtendedDeploymentContext newContext, Events events) {
+        newContext.getAppProps().putAll(this.deploymentContext.getAppProps());
+        newContext.getModulePropsMap().putAll(this.deploymentContext.getModulePropsMap());
+        this.deploymentContext.getModuleMetadata().forEach(newContext::addModuleMetaData);
         this.getDescriptorMetadata()
                 .entrySet()
-                .forEach(e -> context.addTransientAppMetaData(e.getKey(), e.getValue()));
-        this.setDeploymentContext(context);
+                .forEach(e -> newContext.addTransientAppMetaData(e.getKey(), e.getValue()));
+        this.deploymentContext = newContext;
+        this.previousClassLoaders = getClassLoaders(this.applicationInfo);
+        events.send(
+                new EventListener.Event<ApplicationInfo>(
+                        Deployment.APPLICATION_UNLOADED,
+                        this.applicationInfo
+                ),
+                false
+        );
```

Comment:
`APPLICATION_UNLOADED` event snippets will be removed under https://payara.atlassian.net/browse/ECOSYS-90 partial start/stop of the container.